colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import defaultdict
+import csv
+from io import StringIO
+import re
+from typing import Dict, List, NamedTuple, Optional
+
+from colcon_sanitizer_reports._sanitizer_section import SanitizerSection
+
+# The start line of a section can be found with the following regex. Additionally, any prefix that
+# is prepended by the logging system can be extracted and be used to lstrip following section lines.
+_FIND_SECTION_START_LINE_REGEX = \
+    re.compile(r'^(?P<prefix>.*?)(==\d+==|)(WARNING|ERROR):.*Sanitizer:.*$')
```

Comment:
I'm confused by the regexp fragment `(?P<prefix>.*?)`. Match one or more zero or more wildcard characters?

Reply:
It's a non-greedy match. The look-ahead portion is (==\d+==|) and can match either the ==\d+== or nothing. Making the match non-greedy means it will prefer to match as little as possible and make the look-ahead portion match. If it were greedy, the prefix would consume as much as possible and make the next portion match.

As far as the zero-or-more wildcard characters for the prefix, it could match a blank prefix. In that case, the log file would need to have zero interleaving from multi-threaded writers or this code would start to fail.

Reply:
Thanks for clarifying.